theano/scalar/basic.py

```diff
@@ -3807,3 +3812,79 @@ def __setstate__(self, d):
         self.init_fgraph()
         self.init_py_impls()
         assert self._c_code
+
+
+class Compositef32(object):
+    # This is a dict of scalar op classes that need special handling
+    special = {}
+
+    def apply(self, fgraph):
+        mapping = {}
+        topo = fgraph.toposort()
+        for i in fgraph.inputs:
+            if i.dtype == 'float16':
+                mapping[i] = get_scalar_type('float32')()
+            else:
+                mapping[i] = i
+        for node in topo:
+            # Patch up for constants
+            for i in node.inputs:
+                if i not in mapping:
+                    assert type(i) is ScalarConstant
+                    mapping[i] = i
+            if type(node.op) in self.special:
+                self.special[type(node.op)](node, mapping)
+                continue
+            # make sure we won't produce any float16.
+            assert not any(o.dtype == 'float16' for o in
+                           node.op.output_types([mapping[i].type for i in node.inputs]))
+            new_node = node.clone_with_new_inputs(
+                [mapping[i] for i in node.inputs],
+                strict=False)
+            for o, no in zip(node.outputs, new_node.outputs):
+                mapping[o] = no
+
+        new_ins = [mapping[i] for i in fgraph.inputs]
+        new_outs = [mapping[o] for o in fgraph.outputs]
+        return new_ins, new_outs
+
+composite_f32 = Compositef32()
+
+
+def handle_cast(node, mapping):
+    inp = mapping[node.inputs[0]]
+    out = node.outputs[0]
+    node_ok = False
+    if node.op.o_type == float16:
+        if node.inputs[0].type == float32:
+            # cast f32 -> f16, remove
+            mapping[out] = inp
+            return
+        else:
+            # cast to f16, convert to f32
+            new_out = cast(inp, 'float32')
+            # change the node for the following if
+            node = new_out.owner
+            mapping[out] = new_out
+            node_ok = True
+    if node.inputs[0].type == float16:
+        if node.op.o_type == inp.type:
+            # cast f16 to new input type, remove
+            mapping[out] = inp
```

Comment:
The if above checks that we are casting what was previously a float16 input.  The second if checks if the new input type is the same as the casting output type.  In any case, this should always be correct no matter what types are involved.

In practice this should only trigger for cast of float16 to float32.  And in this case we can remove the cast since the input will already be float32.